lib/pbench/server/db/models/tracker.py

```diff
@@ -0,0 +1,546 @@
+                    this is retained if specified, or will be constructed
+                    from "path" if not present.
+                "name": The dataset name (file path basename minus ".tar.xz");
+                    this is retained if specified, or will be constructed from
+                    "path" if not present.
+                "state": The desired state to advance the dataset.
+
+        Raises:
+            DatasetSqlError: problem interacting with Database
+            DatasetNotFound: the specified dataset doesn't exist
+            DatasetTerminalStateViolation: dataset is in terminal state and
+                can't be advanced
+            DatasetBadStateTransition: dataset cannot be advanced to the
+                specified state
+
+        Returns:
+            Dataset: a dataset object in the desired state (if specified)
+        """
+        # Make sure we have controller and name from path
+        Dataset._render_path(kwargs)
+        controller = kwargs.get("controller")
+        name = kwargs.get("name")
+        try:
+            dataset = (
+                Database.db_session.query(Dataset)
+                .filter_by(controller=controller, name=name)
+                .first()
+            )
+        except SQLAlchemyError as e:
+            Dataset.logger.warning(
+                "Error attaching {}>{}: {}", controller, name, str(e)
+            )
+            raise DatasetSqlError(f"Error attaching {controller}:{name}: {e}")
+        else:
+            if dataset is None:
+                Dataset.logger.warning("{}>{} not found", controller, name)
+                raise DatasetNotFound(f"No dataset {controller}:{name}")
+            elif "state" in kwargs:
+                dataset.advance(kwargs.get("state"))
+        return dataset
+
+    def __str__(self):
+        """
+        __str__ Return a string representation of the dataset
+
+        Returns:
+            string: Representation of the dataset
+        """
+        return f"{self.user}>{self.controller}>{self.name}"
+
+    def advance(self, new_state):
+        """
+        advance Modify the state of the Dataset object, if the new_state is
+        a valid transition from the dataset's current state.
+
+        Args:
+            new_state (State ENUM): New desired state for the dataset
+
+        Raises:
+            DatasetTerminalStateViolation: The dataset is in a terminal state
+                that cannot be changed.
+            DatasetBadStateTransition: The dataset does not support transition
+                from the current state to the desired state.
+        """
+        if self.state not in self.transitions:
+            self.logger.error(
+                "Terminal state {} can't advance to {}", self.state, new_state
+            )
+            raise DatasetTerminalStateViolation(
+                f"Dataset state {self.state} is terminal and cannot be advanced"
+            )
```

Comment:
I left it out here because the new state isn't really relevant; you can't advance to _any_ state because of the current state. But, sure. Consistency isn't always evil. ;-)

Reply:
I pointed it out because the new state might offer a clue as to why the error occurred, as to why the code was attempting to transition out of a terminal state.

Consistency is rarely evil when it reduces surprise, and more information in error cases is almost aways better, especially so in cases which should never happen.  :-)